### PR TITLE
Concurrent replace should work with supervisors using concurrent locks

### PR DESCRIPTION
**Pull request by AmatyaAvadhanula**
A concurrent replace should allow the existing pending segments of a peon to be upgraded via the supervisor.
Currently, we check if there is such an active supervisor for streaming ingestion when the context value of `taskLockType`: `APPEND`.
This PR ensures that `useConcurrentLocks`:`true` works as well.

Please note that if `useConcurrentLocks` is set to `false`, it would supersede `taskLockType`:`APPEND`

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java [hunk 1]
```diff
@@ -87,23 +87,32 @@ public Optional<String> getActiveSupervisorIdForDatasourceWithAppendLock(String
...
           );
+
+          if (useConcurrentLocks == null) {
+            useConcurrentLocks = false;
+            taskLockType = QueryContexts.getAsEnum(
+                Tasks.TASK_LOCK_TYPE,
+                context.get(Tasks.TASK_LOCK_TYPE),
+                TaskLockType.class
+            );
+          }
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorManager.java:108]:
```suggestion
          Boolean useConcurrentLocks = QueryContexts.getAsBoolean(
              Tasks.USE_CONCURRENT_LOCKS,
              context.get(Tasks.USE_CONCURRENT_LOCKS)
          );

          if (useConcurrentLocks == null) {
            TaskLockType taskLockType = QueryContexts.getAsEnum(
                Tasks.TASK_LOCK_TYPE,
                context.get(Tasks.TASK_LOCK_TYPE),
                TaskLockType.class
            );
            if (taskLockType == null) {
                hasAppendLock = Tasks.DEFAULT_USE_CONCURRENT_LOCKS;
            } else if (taskLockType == TaskLockType.APPEND) {
                hasAppendLock = true;
            } else {
                hasAppendLock = false;
            }
          } else {
              hasAppendLock = useConcurrentLocks;
          }
```

**Review by kfaraz — COMMENTED** [review]
Left some comments to make the flow easy to understand.

**Review by kfaraz — APPROVED** [review]
Thanks for the fix, @AmatyaAvadhanula !

**Comment by AmatyaAvadhanula** [comment]
Thank you for the review @kfaraz

**Pull request merged**